src/pyclass.jl

```diff
@@ -0,0 +1,164 @@
+
+`def_py_class` creates a Python class. `@pydef` macros expand into a call to
+this function.
+
+Arguments
+---------
+- `methods`: a vector of tuples `(py_name::String, jl_fun::Function)`
+   py_name will be a method of the Python class, which will call `jl_function`
+- `base_classes`: the Python base classes to inherit from.
+
+Return value: the created class (::PyTypeObject)
+"""
+function def_py_class(type_name::AbstractString, methods::Vector;
+                      base_classes=[], getsets::Vector=[])
+    # Only create new-style classes
+    base_classes = union(base_classes, [pybuiltin("object")])
+    new_type = pybuiltin("type")(type_name, tuple(base_classes...), Dict())
+    for (py_name, jl_fun) in methods
+        new_type[py_name::Symbol] = jlfun2pyfun(jl_fun::Function)
+    end
+    for (py_name, getter, setter) in getsets
+        new_type[py_name::Symbol] = pyproperty(jlfun2pyfun(getter),
+                                               jlfun2pyfun(setter))
+    end
+    new_type
+end
+        
+
+######################################################################
+# @pydef macro
+
+
+function parse_pydef(expr)
+    if @capture(expr, begin type type_name_ <: base_classes_expr_
+                    lines__
+                end end)
+        if !@capture(base_classes_expr, (base_classes__,))
+            base_classes = (base_classes_expr,)
+        end
+    else
+        @assert(@capture(expr, type type_name_
+                    lines__
+            end), "Malformed @pydef expression")
+        
+        base_classes = []
+    end
+    if isa(lines[1], Expr) && lines[1].head == :block 
+        # unfortunately, @capture fails to parse the type's fields correctly
+        # It's been reported and fixed, we can remove this line on the next
+        # MacroTools release (> v0.2)
+        lines = lines[1].args
+    end
+
+    function_defs = Expr[] # vector of :(function ...) expressions
+    methods = Tuple{Any, Symbol}[] # (py_name, jl_method_name)
+    getter_dict = Dict{Any, Symbol}() # python_var => jl_getter_name
+    setter_dict = Dict{Any, Symbol}() 
+    method_syms = Dict{Any, Symbol}() # see below
+    for line in lines
+        if !isa(line, LineNumberNode) && line.head != :line # need to skip those
+            @assert line.head == :(=) "Malformed line: $line"
+            lhs, rhs = line.args
+            @assert @capture(lhs,py_f_(args__)) "Malformed left-hand-side: $lhs"
+            if isa(py_f, Symbol)
+                # Method definition
+                # We save the gensym to support multiple dispatch
+                #    readlines(io) = ...
+                #    readlines(io, nlines) = ...
+                # otherwise the first and second `readlines` get different
+                # gensyms, and one of the two gets shadowed by the other.
+                jl_fun_name = get!(method_syms, py_f, gensym(py_f))
+                push!(function_defs, :(function $jl_fun_name($(args...))
```

Comment:
Right here, you could have a check: if `py_f == :init`, then append a `nothing` line to the function body.